subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/SystemDefaultSSLContextFactory.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.resource.transport.http;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+
+import javax.annotation.Nullable;
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.security.KeyStore;
+
+/**
+ * This class contains SSLContext initialization similar to what SSLContext.getDefault() does.
+ * Unfortunately, SSLContext.getDefault() can not be used because it is heavily cached by the JVM,
+ * including keyStore and trustStore managers.
+ * There is no way to reset these caches because they are private and are initialized in static sections.
+ */
+@NonNullApi
+public class SystemDefaultSSLContextFactory {
+    private static final String NONE = "NONE";
+    private static final String P11KEYSTORE = "PKCS11";
+
+    private static final Logger LOGGER = Logging.getLogger(SystemDefaultSSLContextFactory.class);
+
+    static SSLContext create() throws Exception {
+        SSLContext context = SSLContext.getInstance("TLS");
+        context.init(getKeyManagers(), getTrustManagers(), null);
+        return context;
+    }
+
+    private static KeyManager[] getKeyManagers() throws Exception {
+        String keyStorePath = System.getProperty("javax.net.ssl.keyStore", "");
+        String keyStoreType = System.getProperty("javax.net.ssl.keyStoreType", KeyStore.getDefaultType());
+        String keyStoreProvider = System.getProperty("javax.net.ssl.keyStoreProvider", "");
+        String keyStorePasswordString = System.getProperty("javax.net.ssl.keyStorePassword", "");
+
+        if (P11KEYSTORE.equals(keyStoreType) && !NONE.equals(keyStorePath)) {
+            throw new IllegalArgumentException("if keyStoreType is " + P11KEYSTORE + ", then keyStore must be " + NONE);
```

Comment:
💭 No arguments to this method.  Maybe `IllegalStateException` would be more accurate?